rascal-textmate-core/src/main/rascal/util/ListUtil.rsc

```diff
@@ -0,0 +1,33 @@
+module util::ListUtil
+
+import List;
+
+@synopsis{
+    Representation of a traversal direction along a list
+}
+
+data Direction   // Traverse lists...
+    = forward()  //   - ...from left to right;
+    | backward() //   - ...from right to left.
+    ;
+
+@synopsis{
+    Reorder a list according to the specified direction
+}
+
+list[&T] reorder(list[&T] l, forward())  = l;
+list[&T] reorder(list[&T] l, backward()) = reverse(l);
+
+@synopsis{
+    Removes multiple occurrences of elements in a list. The last occurrence
+    remains (cf. `List::dup`).
+}
+
+list[&T] dupLast(list[&T] l) = reverse(dup(reverse(l))); // TODO: Optimize/avoid `reverse`-ing?
+
+@synopsis{
+    Checks if list `l1` is a strict prefix of list `l2`
+}
+
+bool isStrictPrefix(list[&T] l1, list[&T] l2)
+    = size(l1) < size(l2) && !any(i <- [0..size(l1)], l1[i] != l2[i]);
```

Comment:
ps, in the challenge:

```
bool isStrictPrefix(list[&T] l1, list[&T] l2)
    = size(l1) < size(l2) && l1 == l2[..size(l2);
```

is even shorter ;)

Or even this ;)

```
bool isStrictPrefix(list[&T] l1:[_,*_], [*l1, _, *_]) = true;
default bool isStrictPrefix(_, _) = false;
```

Reply:
Benchmarks are hard ;) these are doing to much in the loop that is not part of the focus.

On my machine, I get these numbers:

```
rascal>main_old();
Reloading module Bench
("v2":213,"v3":171,"v4":176,"v5":218)
```

and after rewriting the benchmark:

```
void main(int i = 0) {
    full = [0..1000];
    needle = [0..i];
    rounds = [0..1000];
    cases = (
        "v2": void() { for (_ <- rounds) { isStrictPrefix2(needle, full); } },
        "v3": void() { for (_ <- rounds) { isStrictPrefix3(needle, full); } },
        "v4": void() { for (_ <- rounds) { isStrictPrefix4(needle, full); } },
        "v5": void() { for (_ <- rounds) { isStrictPrefix5(needle, full); } }
    );
    println(benchmark(cases));
}
```

it prints:

```
Reloading module Bench
("v2":71,"v3":47,"v4":37,"v5":24)
```

suddenly v5 is the fastest, and the numbers are all quite a bit lower, let's increase the rounds to 100000.

```
rascal>main()
("v2":908,"v3":852,"v4":839,"v5":187)
```

so yeah, v5 might be a quite a bit faster than the rest of them ;)

Reply:
Thank you, that's interesting! 🙂